### DISCUSSION TITLE
How to add some classes in build time even if they do not exist in the source code?

### DISCUSSION THREAD

**Discussion by ilia4bit** [tailwindlabs/tailwindcss#10192] [Help] [answered]
Hey, I'm new to TailwindCSS. I have a Django (Python web framework) project and I decide to use tailwind as my styling framework. But I realize that when I execute the `build` command, only classes explicitly written in the source code will exist in the `output.css` file. But I want to give users some customization options. I will give you a simplified example:

‍‍‍models.py (where you define database schema)
```python
class Grid:
 cols_number: int = IntegerField(min=1, max=12) # to generate grid-cols-{n}
 rows_number: int = IntegerField(min=1, max=12) # to generate grid-cols-{n}
```
views.py (where you handle request and response and bridge between models and template layer)
```python
# this python function gets Request a parameter and returns an HTTP response with context
def grid_view(request):
 grid = Grid.objects.get(id=2) # selects from grid table where id = 2
 context = {"cols_number": grid.cols_number, "rows_number: grid.rows_number }
 return HTTPResponse(request=request, template_name="grid.html", context=context)
```

grid.html (template file based on Django template engine
context values which returned by gird_view accessible with {{ value }} syntax
```html
<div class="grid grid-cols-{{ col }} grid-rows-{{ row }}>
 <!-- some elements -->
</div>
```
as you see in this example there aren't grid-cols- and grid-rows- classes in TailwindCSS, so How do I tell Tailwindcss to add grid utility classes even if they don't exist explicitly in the source code

**Comment by wongjn** [comment, marked as answer]
You have several options:

## Use the `style` attribute

```html
<div class="grid" style="grid-template-columns: repeat({{ col }}, 1fr); grid-template-rows: repeat({{ row }}, 1fr)">
```

## Use CSS variables

```html
<div class="grid grid-cols-[var(--cols)] grid-rows-[var(--rows)]" style="--cols: {{ col }}; --rows: {{ row }}">
```

## Use [`safelist`](https://tailwindcss.com/docs/content-configuration#safelisting-classes)

```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  // …
  safelist: [
    // Generates:
    // 'grid-cols-1',
    // 'grid-rows-1',
    // 'grid-cols-2',
    // 'grid-rows-2',
    // …
    ...Array(12).fill(null).flatMap(i => [`grid-cols-${i + 1}`, `grid-rows-${i + 1}`),
  ],
  // …
};
```

---

You also need to ensure that you have values for `gridTemplateRows` up to 12 since by default, [it only goes up to 6](https://tailwindcss.com/docs/grid-template-rows):
```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  // …
  theme: {
    extend: {
      // Generates:
      // 7: 'repeat(7, 1fr)',
      // 8: 'repeat(8, 1fr)',
      // …
      gridTemplateRows: Object.fromEntries(
        Array(6).fill(null).map(i => [i + 7, `repeat(${i + 7}, 1fr)`]),
      ),
      // …
    },
  },
  // …
};
```

**Reply by ilia4bit** [reply]
Thanks a lot 🙏🏼.
